File: spring_java_orm_sample/src/main/java/com/example/demo/mybatis/converter/BookOperationConverterMybatis.java
package com.example.demo.mybatis.converter;

import com.example.demo.api.response.BookResponse;
import com.example.demo.api.response.BookStockResponse;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Component;

import com.example.demo.mybatis.entity.BookWithPublisherName;

import java.util.List;

@Component
@RequiredArgsConstructor
public class BookOperationConverterMybatis {
    private final ModelMapper modelMapper;

    public BookResponse toResponse(BookWithPublisherName book) {
        final var response = modelMapper.map(book, BookResponse.class);
        response.setBookStockList(book.getBookStockList().stream()
            .map(bookStock -> modelMapper.map(bookStock, BookStockResponse.class))
            .toList());
        return response;
    }

    public List<BookResponse> toResponse(List<BookWithPublisherName> books) {
        return books.stream().map(this::toResponse).toList();
    }
}

File: codex_poting/kotlin/src/main/kotlin/com/example/demo/mybatis/converter/BookOperationConverterMybatis.kt
package com.example.demo.mybatis.converter

import com.example.demo.api.response.BookResponse
import com.example.demo.api.response.BookStockResponse
import com.example.demo.mybatis.entity.BookStockWithStoreName
import com.example.demo.mybatis.entity.BookWithPublisherName
import org.modelmapper.ModelMapper
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component

@Component
@Profile("mybatis")
class BookOperationConverterMybatis(private val modelMapper: ModelMapper) {
    fun toResponse(book: BookWithPublisherName): BookResponse {
        val response = modelMapper.map<BookResponse>(book, BookResponse::class.java)
        response.bookStockList = book.bookStockList!!.filterNotNull()
            .map { bookStock: BookStockWithStoreName ->
                modelMapper.map<BookStockResponse>(
                    bookStock,
                    BookStockResponse::class.java
                )
            }
        return response
    }

    fun toResponse(books: MutableList<BookWithPublisherName?>): MutableList<BookResponse> {
        return books.filterNotNull().map { toResponse(it) }.toMutableList()
    }
}
